README.md
node-perceptualdiff
===================

A node port of the perceptualdiff image comparison (http://pdiff.sourceforge.net) with some additional features


#Usage:

The package can be used in two different ways:
 * per command line; just as the original project
 * through a class in your code


##Command-Line usage:

The command-line tool can be found in the ```bin``` directory. You can run the application with

```
node ./bin/perceptualdiff.js <image1>.png <image2>.png
```
Use ```image1``` and ```image2``` as the images you want to compare.

**Note:** This port only supports PNGs!

The command-line tool exposes a couple of flags and parameters for the comparison:
```
--verbose           Turn on verbose mode"
--fov deg           Field of view in degrees [0.1, 89.9] (default: 45.0)
--threshold p       Number of pixels/percent p below which differences are ignored
--threshold-image p Number of pixels/percent p below which differences are not generated (see --output)
--threshold-type p  'pixel' and 'percent' as type of threshold. (default: pixel)
--pyramid-levels p  Number of levels of Laplacian pyramids. (default: 3)
--gamma g           Value to convert rgb into linear space (default: 2.2)
--luminance l       White luminance (default: 100.0 cdm^-2)
--luminance-only    Only consider luminance; ignore chroma (color) in the comparison
--color-factor      How much of color to use [0.0, 1.0] (default: 1.0)
--scale             Scale images to match each others dimensions
--sum-errors        Print a sum of the luminance and color differences
--output o          Write difference to the file o
--version           Print version
--help              This help
```

Most of the parameters were exposed as the original project does. However, I changed a couple of parameter signatures
to make the interface a little bit more consistent.
* ```--luminanceonly``` was renamed to ```luminance-only```
* ```--colorfactor``` was renamed to ```color-factor```

Since the PNG library, I use, does not support resampling, I needed to remove this feature for now.
So, there is no ```--resample```. Please resample the images through other means before using this diff-tool.

I also added a couple additional features and some were exposed to the command-line tool:
* ```--threshold-image p``` makes it possible to skip some of the comparison, reducing the time spent analysing the
images as node is a LOT slower than C is. This feature will also skip the creation of output images if this threshold is not reached; it simply stops caring if the difference is below the threshold.
* ```--threshold-type p``` changes the threshold by considering absolute pixels or percentage of total pixels. The values are ```pixel``` and ```percent``` respectively.
* ```--pyramid-levels p``` specifies the detail of the comparison - the higher the number is, the higher the comparison resolution but also the longer it will take. ```2``` is the lowest number possible. The original perceptualdiff tool used internally 8 as default. Again, node is just too slow.


##Class usage:

The package can also be used directly in code, without going through the command-line.

**Example:**
```JavaScript
var PerceptualDiff = require('perceptualdiff');

var diff = new PerceptualDiff({
    imageAPath: '...',
    imageBPath: '...',

    scale: true,
    verbose: true,
    pyramidLevels: 5,

    thresholdType: PerceptualDiff.THRESHOLD_PERCENT,
    threshold: 0.01,
    imageThreshold: 0.005,

    outputPath: '...'
});

diff.run(function (passed) {
    console.log(passed ? 'Passed' : 'Failed');
});
```

All the parameters that were available in the command-line tool are also available through the class constructor - they use
camelCasing instead of snake-casing. The class exposes additional parameters that are not available from the command-line.
* ```imageAPath``` Defines the path to the first image that should be compared (required)
* ```imageBPath``` Defines the path to the second image that should be compared (required)
* ```imageOutputPath``` Defines the path to the output-file. If you leaves this one off, then this feature is turned-off.
* ```verbose``` Verbose output (default: false)
* ```luminanceOnly``` Only consider luminance; ignore chroma (color) in the comparison (default: false)
* ```sumErrors``` Print a sum of the luminance and color differences (default: false)
* ```fieldOfView``` Field of view in degrees [0.1, 89.9] (default: 45.0)
* ```gamma``` Value to convert rgb into linear space (default: 2.2)
* ```luminance``` White luminance (default: 100.0 cdm^-2)
* ```thresholdType``` Type of threshold check. This can be PerceptualDiff.THRESHOLD_PIXEL and PerceptualDiff.THRESHOLD_PERCENT (default: THRESHOLD_PIXEL)
* ```threshold``` Number of pixels/percent p below which differences are ignored (default: 100)
* ```imageThreshold``` Number of pixels/percent p below which differences are not generated (default: 50)
* ```colorFactor``` How much of color to use [0.0, 1.0] (default: 1.0)
* ```pyramidLevels``` Number of levels of Laplacian pyramids. (default: 3)
* ```scale``` Scale images to match each others dimensions (default: false)
* ```outputMaskRed``` Red intensity for the difference highlighting in the output file (default: 255 - full red)
* ```outputMaskGreen``` Green intensity for the difference highlighting in the output file (default: 0)
* ```outputMaskBlue``` Blue intensity for the difference highlighting in the output file (default: 0)
* ```outputMaskAlpha``` Alpha intensity for the difference highlighting in the output file (default: 255)
* ```outputMaskOpacity``` Opacity of the pixel for the difference highlighting in the output file (default: 0.7 - slightly transparent)
* ```outputBackgroundRed``` Red intensity for the background in the output file (default: 0)
* ```outputBackgroundGreen``` Green intensity for the background in the output file (default: 0)
* ```outputBackgroundBlue``` Blue intensity for the background in the output file (default: 0)
* ```outputBackgroundAlpha``` Alpha intensity for the background in the output file (default: 0 - transparent)
* ```copyImageAToOutput``` Copies the first image to the output image before the comparison begins. This will make sure that the output image will highlight the differences on the first image.
* ```copyImageBToOutput``` Copies the second image to the output image before the comparison begins. This will make sure that the output image will highlight the differences on the second image.


###Logging:

By default, the logger doesn't log events anywhere, but you can overwrite this behavior by overwriting ```PerceptualDiff.log```:

```JavaScript
var diff = new PerceptualDiff({
    ...
});

diff.log = function (text) {
    // Do whatever you want to do
};

...
```


#Example:

There are some examples in the ```examples``` folder, where I used screenshots of Wikipedia to check for visual regressions.
You can find examples for:
* Missing DOM elements in ```hidden_regression```
* Disrupted sorting in ```sorting_regression```
* Color changes in ```style_regression```
* Text capitalization in ```text_regression```

All screenshots were compared to ```wikipedia_approved.png```, a previously approved screenshot without a regression.
Each of the regression has the screenshot and the output result, highlighting the differences.


#TODO:

* Code documentation


#LICENSE

The original project was release with the GPL v2 license.

CHANGELOG.md
CHANGELOG
=========

v1.4.1
* Add editor-config file
* Add change-log
* Bump versions

v1.4.0 - Clean-up 02/26/15

v1.3.0 - Initial release 08/24/14


lib/metric.js
/*
 Metric
 Copyright (C) 2006-2011 Yangli Hector Yee
 Copyright (C) 2011-2014 Steven Myint
 Copyright (C) 2014-2015 Marcel Erz

 This program is free software; you can redistribute it and/or modify it under
 the terms of the GNU General Public License as published by the Free Software
 Foundation; either version 2 of the License, or (at your option) any later
 version.

 This program is distributed in the hope that it will be useful, but WITHOUT ANY
 WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
 PARTICULAR PURPOSE.  See the GNU General Public License for more details.

 You should have received a copy of the GNU General Public License along with
 this program; if not, write to the Free Software Foundation, Inc., 59 Temple
 Place, Suite 330, Boston, MA 02111-1307 USA
 */

function log10 (x) {
	return Math.log(x) / Math.LN10;
}

/*
 * Given the adaptation luminance, this function returns the
 * threshold of visibility in cd per m^2
 * TVI means Threshold vs Intensity function
 * This version comes from Ward Larson Siggraph 1997
 */
function tvi (adaptation_luminance) {
	// returns the threshold luminance given the adaptation luminance
	// units are candelas per meter squared
	var log_a = log10(adaptation_luminance);

	var r;
	if (log_a < -3.94) {
		r = -2.86;
	} else if (log_a < -1.44) {
		r = Math.pow(0.405 * log_a + 1.6, 2.18) - 2.86;
	} else if (log_a < -0.0184) {
		r = log_a - 0.395;
	} else if (log_a < 1.9) {
		r = Math.pow(0.249 * log_a + 0.65, 2.7) - 0.72;
	} else {
		r = log_a - 1.255;
	}

	return Math.pow(10.0, r);
}

// computes the contrast sensitivity function (Barten SPIE 1989)
// given the cycles per degree (cpd) and luminance (lum)
function csf (cpd, lum) {
	var a = 440.0 * Math.pow((1.0 + 0.7 / lum), -0.2);
	var b = 0.3 * Math.pow((1.0 + 100.0 / lum), 0.15);
	return csfCache = a * cpd * Math.exp(-b * cpd) * Math.sqrt(1.0 + 0.06 * Math.exp(b * cpd));
}

/*
 * Visual Masking Function from Daly 1993
 */
function mask (contrast) {
	var a = Math.pow(392.498 * contrast, 0.7);
	var b = Math.pow(0.0153 * a, 4.0);
	return Math.pow(1.0 + b, 0.25);
}

// convert Adobe RGB (1998) with reference white D65 to XYZ
function adobe_rgb_to_xyz (r, g, b) {
	// matrix is from http://www.brucelindbloom.com/
	var x = r * 0.576700 + g * 0.185556 + b * 0.188212;
	var y = r * 0.297361 + g * 0.627355 + b * 0.0752847;
	var z = r * 0.0270328 + g * 0.0706879 + b * 0.991248;

	return {
		x: x, y: y, z: z
	};
}

var global_white = adobe_rgb_to_xyz(1, 1, 1);
var epsilon = 216 / 24389;
var kappa = 24389 / 27;

// convert Adobe RGB (1998) with reference white D65 to XYZ
function adobe_rgb_to_lab (cr, cg, cb) {
	var p = adobe_rgb_to_xyz(cr, cg, cb);
	var x = p.x, y = p.y, z = p.z;

	var f = [];
	var r = [];

	r[0] = x / global_white.x;
	r[1] = y / global_white.y;
	r[2] = z / global_white.z;

	for (var i = 0; i < 3; i++) {
		if (r[i] > epsilon) {
			f[i] = Math.pow(r[i], 1 / 3);
		} else {
			f[i] = (kappa * r[i] + 16) / 116;
		}
	}
	var A = 500 * (f[0] - f[1]);
	var B = 200 * (f[1] - f[2]);

	return {
		A: A, B: B, y: y
	};
}

function adaptation (levels, num_one_degree_pixels) {
	var num_pixels = 1.0;
	var adaptation_level = 0;
	for (var i = 0; i < levels; i++) {
		adaptation_level = i;
		if (num_pixels > num_one_degree_pixels) {
			break;
		}
		num_pixels *= 2;
	}
	// LCOV_EXCL_LINE
	return adaptation_level;
}

module.exports = {
	tvi: tvi,
	csf: csf,
	mask: mask,
	adobe_rgb_to_xyz: adobe_rgb_to_xyz,
	adobe_rgb_to_lab: adobe_rgb_to_lab,
	adaptation: adaptation
};


lib/lpyramid.js
/*
 Laplacian Pyramid
 Copyright (C) 2006-2011 Yangli Hector Yee
 Copyright (C) 2011-2014 Steven Myint
 Copyright (C) 2014-2015 Marcel Erz

 This program is free software; you can redistribute it and/or modify it under
 the terms of the GNU General Public License as published by the Free Software
 Foundation; either version 2 of the License, or (at your option) any later
 version.

 This program is distributed in the hope that it will be useful, but WITHOUT ANY
 WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
 PARTICULAR PURPOSE. See the GNU General Public License for more details.

 You should have received a copy of the GNU General Public License along with
 this program; if not, write to the Free Software Foundation, Inc., 59 Temple
 Place, Suite 330, Boston, MA 02111-1307 USA
 */

function LPyramid (image, width, height, levels) {

	var levelList = this.levelList = [];
	var dim = width * height;
	var j, buf;

	this.width = width;
	this.weight = height;
	this.levels = levels;

	// Make the Laplacian pyramid by successively
	// copying the earlier levels and blurring them
	for (var i = 0; i < this.levels; i++) {
		if (i == 0 || dim <= 1) {
			buf = [];
			for (j = 0; j < dim; j++) {
				buf[j] = image[j];
			}
			levelList[i] = buf;
		} else {
			levelList[i] = [];
			this.convolve(levelList[i], levelList[i - 1]);
		}
	}
}


LPyramid.prototype = {

	// Convolves image b with the filter kernel and stores it in a.
	convolve: function (a, b) {
		var Kernel = [0.05, 0.25, 0.4, 0.25, 0.05];

		var width = this.getWidth();
		var weight = this.getWeight();

		//#pragma omp parallel for
		for (var y = 0; y < weight; y++) {
			for (var x = 0; x < width; x++) {

				var index = y * width + x;
				a[index] = 0;

				for (var i = -2; i <= 2; i++) {
					for (var j = -2; j <= 2; j++) {

						var nx = x + i;
						var ny = y + j;

						if (nx < 0) nx = -nx;
						if (ny < 0) ny = -ny;
						if (nx >= width) nx = (width << 1) - nx - 1;
						if (ny >= weight) ny = (weight << 1) - ny - 1;

						a[index] += Kernel[i + 2] * Kernel[j + 2] * b[ny * width + nx];
					}
				}
			}
		}
	},

	getValue: function (idx, level) {
		return this.levelList[level][idx];
	},

	getLevels: function () {
		return this.levels;
	},

	getWidth: function () {
		return this.width;
	},

	getWeight: function () {
		return this.weight;
	}
};

module.exports = LPyramid;


perceptualdiff.js
/*
 Perceptual Diff
 Copyright (C) 2006-2011 Yangli Hector Yee
 Copyright (C) 2011-2014 Steven Myint
 Ported to Node: Marcel Erz

 This program is free software; you can redistribute it and/or modify it under
 the terms of the GNU General Public License as published by the Free Software
 Foundation; either version 2 of the License, or (at your option) any later
 version.

 This program is distributed in the hope that it will be useful, but WITHOUT ANY
 WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
 PARTICULAR PURPOSE.  See the GNU General Public License for more details.

 You should have received a copy of the GNU General Public License along with
 this program; if not, write to the Free Software Foundation, Inc., 59 Temple
 Place, Suite 330, Boston, MA 02111-1307 USA
 */

var assert = require('assert'),
  LPyramid = require('./lib/lpyramid.js'),
  PNGImage = require('png-image'),
  metrics = require('./lib/metric.js');

function PerceptualDiff (options) {

    this.imageA = null;
    this.imageAPath = options.imageAPath;
    assert.ok(options.imageAPath, "Path to image A not given.");

    this.imageB = null;
    this.imageBPath = options.imageBPath;
    assert.ok(options.imageBPath, "Path to image B not given.");

    this.imageOutput = null;
    this.imageOutputPath = options.imageOutputPath;

    this.verbose = options.verbose || false;

    // Only consider luminance; ignore chroma channels in the comparison.
    this.luminanceOnly = options.luminanceOnly || false;

    // Print a sum of the luminance and color differences of each pixel.
    this.sumErrors = options.sumErrors || false;

    // Field of view in degrees.
    this.fieldOfView = options.fieldOfView || 45;

    // The gamma to convert to linear color space
    this.gamma = options.gamma || 2.2;

    this.luminance = options.luminance || 100;

    // Pixel of Percent
    this.thresholdType = options.thresholdType || PerceptualDiff.THRESHOLD_PIXEL;

    // How many pixels different to ignore.
    this.threshold = options.threshold || 100;

    // How many pixels different to not create image.
    this.imageThreshold = options.imageThreshold || 50;

    // How much color to use in the metric.
    // 0.0 is the same as luminance_only_ = true,
    // 1.0 means full strength.
    this.colorFactor = options.colorFactor || 1.0;

    this.pyramidLevels = options.pyramidLevels || 3;

    this.scale = options.scale || false;

    this.outputMaskRed = options.outputMaskRed || 255;
    this.outputMaskGreen = options.outputMaskGreen || 0;
    this.outputMaskBlue = options.outputMaskBlue || 0;
    this.outputMaskAlpha = options.outputMaskAlpha || 255;

    this.outputBackgroundRed = options.outputBackgroundRed;
    this.outputBackgroundGreen = options.outputBackgroundGreen;
    this.outputBackgroundBlue = options.outputBackgroundBlue;
    this.outputBackgroundAlpha = options.outputBackgroundAlpha;

    this.copyImageAToOutput = options.copyImageAToOutput || false;
    this.copyImageBToOutput = options.copyImageBToOutput || false;

    this.outputMaskOpacity = options.outputMaskOpacity || 0.7;

    this.quick = options.quick || false;
}

PerceptualDiff.THRESHOLD_PIXEL = 'pixel';
PerceptualDiff.THRESHOLD_PERCENT = 'percent';

PerceptualDiff.RESULT_FAILED_DIMENSIONS = 0;
PerceptualDiff.RESULT_FAILED_DIFFERENT = 1;
PerceptualDiff.RESULT_IDENTICAL = 5;
PerceptualDiff.RESULT_INDISTINGUISHABLE = 6;
PerceptualDiff.RESULT_SIMILAR = 7;

PerceptualDiff.prototype = {

    run: function (fn) {
        var self = this;

        PNGImage.log = function (text) {
            self.log('ERROR: ' + text);
            throw new Error('ERROR: ' + text);
        };

        this.imageA = PNGImage.readImage(this.imageAPath, function () {
            self.imageB = PNGImage.readImage(self.imageBPath, function () {

                if (self.scale) self._scale();

                self.imageOutput = PNGImage.createImage(self.imageA.getWidth(), self.imageA.getHeight());

                if (self.copyImageAToOutput) {
                    self.imageA.getImage().bitblt(self.imageOutput.getImage(), 0, 0, self.imageA.getWidth(), self.imageA.getHeight(), 0, 0);

                } else if (self.copyImageBToOutput) {
                    self.imageB.getImage().bitblt(self.imageOutput.getImage(), 0, 0, self.imageB.getWidth(), self.imageB.getHeight(), 0, 0);
                }

                self._yee_compare(fn);
            });
        });
    },

    isPassed: function (result) {
        return (result !== PerceptualDiff.RESULT_FAILED_DIFFERENT) && (result !== PerceptualDiff.RESULT_FAILED_DIMENSIONS);
    },

    hasOutput: function (result) {
        return (result === PerceptualDiff.RESULT_FAILED_DIFFERENT) || (result === PerceptualDiff.RESULT_SIMILAR);
    },

    _scale: function () {

        var min_width, min_height;

        if (this.imageA.getWidth() != this.imageB.getWidth() || this.imageA.getHeight() != this.imageB.getHeight()) {
            min_width = this.imageA.getWidth();
            if (this.imageB.getWidth() < min_width) {
                min_width = this.imageB.getWidth();
            }

            min_height = this.imageA.getHeight();
            if (this.imageB.getHeight() < min_height) {
                min_height = this.imageB.getHeight();
            }

            if (this.verbose) this.log("Scaling to " + min_width + " x " + min_height);

            this.imageA.clip(0, 0, min_width, min_height);
            this.imageB.clip(0, 0, min_width, min_height);
        }
    },

    isAboveThreshold: function (items, total) {
        if ((this.thresholdType === PerceptualDiff.THRESHOLD_PIXEL) && (this.threshold <= items)) {
            return true;
        } else if (this.threshold <= (items / total)) {
            return true;
        }
        return false;
    },

    isAboveImageThreshold: function (items, total) {
        if ((this.thresholdType === PerceptualDiff.THRESHOLD_PIXEL) && (this.imageThreshold <= items)) {
            return true;
        } else if (this.imageThreshold <= (items / total)) {
            return true;
        }
        return false;
    },

    getInfo: function () {
        var thresholdType = (this.thresholdType === PerceptualDiff.THRESHOLD_PIXEL ? 'px' : '%'),
            result = [];

        result.push("Field of view: " + this.fieldOfView + " degrees");
        result.push("Threshold: " + this.threshold + thresholdType);
        result.push("Image threshold: " + this.imageThreshold + thresholdType);
        result.push("Gamma: " + this.gamma);
        result.push("Luminance: " + this.luminance + " cd/m^2");
        result.push("Pyramid-Levels: " + this.pyramidLevels);

        return result;
    },


    log: function (text) {
        // Nothing here; Overwrite this to add some functionality
    },


    _quickIdentical: function (dim) {
        var diff = 0, idx, i,
            outputImageHasImage = this.copyImageAToOutput || this.copyImageBToOutput,
            outputOpacity = outputImageHasImage ? this.outputMaskOpacity : undefined;

        for (i = 0; i < dim; i++) {
            idx = i << 2;

            if (this.imageA.getAtIndex(idx) !== this.imageB.getAtIndex(idx)) {

                diff++;

                if (this.quick) {
                    this.imageOutput.setAtIndex(idx, this.outputMaskRed, this.outputMaskGreen, this.outputMaskBlue, this.outputMaskAlpha, outputOpacity);
                } else if (this.isAboveImageThreshold(diff, dim)) {
                    break;
                }

            } else if (this.quick) {
                this.imageOutput.setAtIndex(idx, this.outputBackgroundRed, this.outputBackgroundGreen, this.outputBackgroundBlue, this.outputBackgroundAlpha);
            }
        }

        return diff;
    },

    _yee_compare: function (fn) {
        var i, ii, resultA, resultB, r, g, b, w, h, dim, diffPixel, a_lum, b_lum, ab, da, db, la, lb,
            outputImageHasImage = this.copyImageAToOutput || this.copyImageBToOutput,
            outputOpacity = outputImageHasImage ? this.outputMaskOpacity : undefined;;

        if ((this.imageA.getWidth() != this.imageB.getWidth()) || (this.imageA.getHeight() != this.imageB.getHeight())) {
            this.log("Image dimensions do not match");
            fn(PerceptualDiff.RESULT_FAILED_DIMENSIONS);
            return;
        }

        w = this.imageA.getWidth();
        h = this.imageA.getHeight();
        dim = w * h;

        diffPixel = this._quickIdentical(dim);
        if (diffPixel == 0) {
            this.log("Images are binary identical");
            fn(PerceptualDiff.RESULT_IDENTICAL);
            return;
        }
        if (!this.isAboveImageThreshold(diffPixel, dim)) {
            this.log("Images are perceptually indistinguishable");
            fn(PerceptualDiff.RESULT_INDISTINGUISHABLE);
            return;
        }

        if (!this.quick) {
            // Assuming colorspaces are in Adobe RGB (1998) convert to XYZ.
            a_lum = [];
            b_lum = [];

            ab = [];

            if (this.verbose) this.log("Converting RGB to XYZ");

            //#pragma omp parallel for
            for (i = 0; i < dim; i++) {
                ii = i << 2;

                r = Math.pow(this.imageA.getRed(ii) / 255, this.gamma);
                g = Math.pow(this.imageA.getGreen(ii) / 255, this.gamma);
                b = Math.pow(this.imageA.getBlue(ii) / 255, this.gamma);
                resultA = metrics.adobe_rgb_to_lab(r, g, b);

                r = Math.pow(this.imageB.getRed(ii) / 255, this.gamma);
                g = Math.pow(this.imageB.getGreen(ii) / 255, this.gamma);
                b = Math.pow(this.imageB.getBlue(ii) / 255, this.gamma);
                resultB = metrics.adobe_rgb_to_lab(r, g, b);

                a_lum[i] = resultA.y * this.luminance;
                b_lum[i] = resultB.y * this.luminance;

                da = resultA.A - resultB.A;
                db = resultA.B - resultB.B;
                da = da * da;
                db = db * db;
                ab[i] = da + db;
            }

            if (this.verbose) this.log("Constructing Laplacian Pyramids");

            la = new LPyramid(a_lum, w, h, this.pyramidLevels);
            lb = new LPyramid(b_lum, w, h, this.pyramidLevels);

            var num_one_degree_pixels = 2 * Math.tan(this.fieldOfView * 0.5 * Math.PI / 180) * 180 / Math.PI;
            var pixels_per_degree = w / num_one_degree_pixels;

            if (this.verbose) this.log("Performing test");

            var adaptation_level = metrics.adaptation(this.pyramidLevels, num_one_degree_pixels);

            var cpd = [];
            cpd[0] = 0.5 * pixels_per_degree;
            for (i = 1; i < this.pyramidLevels; i++) {
                cpd[i] = 0.5 * cpd[i - 1];
            }
            var csf_max = metrics.csf(3.248, 100);

            var F_freq = [];
            for (i = 0; i < this.pyramidLevels - 2; i++) {
                F_freq[i] = csf_max / metrics.csf(cpd[i], 100);
            }

            diffPixel = 0;
            var error_sum = 0;

            var contrast = [];
            var F_mask = [];

            //#pragma omp parallel for reduction(+ : diffPixel) reduction(+ : error_sum)
            var index;
            for (index = 0; index < dim; index++) {
                var sumContrast = 0;
                for (i = 0; i < this.pyramidLevels - 2; i++) {
                    var n1 = Math.abs(la.getValue(index, i) - la.getValue(index, i + 1));
                    var n2 = Math.abs(lb.getValue(index, i) - lb.getValue(index, i + 1));
                    var numerator = (n1 > n2) ? n1 : n2;
                    var d1 = Math.abs(la.getValue(index, i + 2));
                    var d2 = Math.abs(lb.getValue(index, i + 2));
                    var denominator = (d1 > d2) ? d1 : d2;
                    if (denominator < 1e-5) denominator = 1e-5;
                    contrast[i] = numerator / denominator;
                    sumContrast += contrast[i];
                }
                if (sumContrast < 1e-5) sumContrast = 1e-5;

                var adapt = la.getValue(index, adaptation_level) + lb.getValue(index, adaptation_level) * 0.5;
                if (adapt < 1e-5) adapt = 1e-5;
                for (i = 0; i < this.pyramidLevels - 2; i++) {
                    F_mask[i] = metrics.mask(contrast[i] * metrics.csf(cpd[i], adapt));
                }

                var factor = 0;
                for (i = 0; i < this.pyramidLevels - 2; i++) {
                    factor += contrast[i] * F_freq[i] * F_mask[i] / sumContrast;
                }
                if (factor < 1) factor = 1;
                if (factor > 10) factor = 10;

                var delta = Math.abs(la.getValue(index, 0) - lb.getValue(index, 0));
                error_sum += delta;

                var pass = true;

                // pure luminance test
                if (delta > factor * metrics.tvi(adapt)) pass = false;

                if (pass && !this.luminanceOnly) {
                    // CIE delta E test with modifications
                    var color_scale = this.colorFactor;
                    // ramp down the color test in scotopic regions
                    if (adapt < 10) {
                        // Don't do color test at all.
                        color_scale = 0;
                    }
                    var delta_e = ab[index] * color_scale;
                    error_sum += delta_e;
                    if (delta_e > factor) pass = false;
                }

                if (!pass) {
                    diffPixel++;
                    this.imageOutput.setAtIndex(index << 2, this.outputMaskRed, this.outputMaskGreen, this.outputMaskBlue, this.outputMaskAlpha, outputOpacity);

                } else {
                    this.imageOutput.setAtIndex(index << 2, this.outputBackgroundRed, this.outputBackgroundGreen, this.outputBackgroundBlue, this.outputBackgroundAlpha);
                }
            }
        }

        var self = this;
        var completion = function () {
            if (self.isAboveThreshold(diffPixel, dim)) {
                self.log("Images are visibly different");
                self.log(diffPixel + " pixels are different");
                if (self.sumErrors) self.log(error_sum + " error sum");
                fn(PerceptualDiff.RESULT_FAILED_DIFFERENT);
            } else {
                self.log("Images are similar");
                self.log(diffPixel + " pixels are different");
                fn(PerceptualDiff.RESULT_SIMILAR);
            }
        };

        // Always output image difference if requested.
        if (this.imageOutputPath) {
            this.imageOutput.writeImage(this.imageOutputPath, function () {
                self.log("Wrote difference image to " + self.imageOutputPath);
                completion();
            });
        } else {
            completion();
        }
    }
};

PerceptualDiff.version = "1.3.15";

module.exports = PerceptualDiff;
